File: backend/app/schemas/gm.py
"""GM API Schema（M1 + M2 development）。"""

from __future__ import annotations

from pydantic import BaseModel, Field, model_validator


class GmSetCharacterRequest(BaseModel):
    """``POST /gm/character/set``：字段皆可选，至少一项。"""

    major_realm: str | None = None
    realm_stage: int | None = Field(default=None, ge=1)
    cultivation_points: int | None = Field(default=None, ge=0)
    realm_progress: int | None = Field(default=None, ge=0)
    body_tempering_points: int | None = Field(default=None, ge=0)
    crafting_exp: int | None = Field(default=None, ge=0)
    breakthrough_grade: str | None = None
    membership_tier: str | None = None
    spirit_stones: int | None = Field(default=None, ge=0)
    idle_direction: str | None = None
    status: str | None = None
    clear_offline_pending: bool | None = None
    # --- M3 战斗成型调试字段 ---
    set_stamina: int | None = Field(default=None, ge=0, description="直接设置体力值")
    trial_puppet_count: int | None = Field(
        default=None,
        ge=0,
        description="设置试炼木傀持有数",
    )
    reset_snapshot_cooldown: bool | None = Field(
        default=None,
        description="清除快照手动更新冷却",
    )
    force_refresh_snapshot: bool | None = Field(
        default=None,
        description="立即用当前防守预设重建快照",
    )
    # --- M4 双线程成长调试 ---
    divine_sense_capacity_bonus: int | None = Field(default=None, ge=0)
    array_craft_level: int | None = Field(default=None, ge=0)
    force_jindan: bool | None = Field(default=None, description="一键设为金丹初期")
    grant_craft_materials: bool | None = Field(default=None, description="发放配方测试材料")
    grant_test_pet: bool | None = Field(default=None, description="发放测试灵宠")
    clear_craft_jobs: bool | None = Field(default=None, description="清空工坊队列")
    clear_divine_sense_backlash: bool | None = Field(default=None)
    # --- M5 环境与轮回 ---
    force_shichen: str | None = Field(default=None, description="强制当前时辰 id")
    force_weather: str | None = Field(default=None, description="强制天气 id")
    start_tribulation: bool | None = Field(default=None, description="强制进入渡劫准备")
    set_awaiting_ferry: bool | None = Field(default=None, description="直接置待引渡")
    force_ferry_timeout: bool | None = Field(default=None, description="立即超时强制轮回")
    mark_story_node: str | None = Field(default=None, description="标记已历剧情节点")
    fate_luck: int | None = Field(default=None, description="设置气运")
    demonic_nature: int | None = Field(default=None, description="设置魔性")
    force_yuanying_peak: bool | None = Field(
        default=None,
        description="一键设为元婴大圆满并灌满进度",
    )
    spirit_root_tags: list[str] | None = Field(
        default=None,
        description="设置灵根环境标签（如 thunder_root）",
    )
    force_tribulation_outcome: str | None = Field(
        default=None,
        description="强制渡劫结局：won / failed / fallen（验收）",
    )
    grant_acceptance_constitution: bool | None = Field(
        default=None,
        description="发放验收体质主/副词条并自动镶嵌",
    )
    # --- M6 大道 / 道主 ---
    force_true_immortal: bool | None = Field(
        default=None,
        description="一键设为真仙初期（可开道）",
    )
    grant_dao_pool: list[str] | None = Field(
        default=None,
        description="灌入指定道 id 入道池",
    )
    set_dao_lord: str | None = Field(
        default=None,
        description="强制任命为该道道主；空字符串清空自己的道主身份",
    )
    open_dao_challenge_window: bool | None = Field(
        default=None,
        description="强制当前处于道主开窗（进程标志；赛会日程仍以 contest 为准）",
    )
    open_dao_contest_now: bool | None = Field(
        default=None,
        description="立刻开赛：关闭报名并收口本场道主之争（P1）",
    )
    clear_dao_challenge_cooldown: bool | None = Field(
        default=None,
        description="清空挑战冷却",
    )
    push_world_env: bool | None = Field(
        default=None,
        description="经 WS 推一条 world.env（若 Hub 有连接）",
    )
    set_dao_qi: int | None = Field(default=None, ge=0, description="直接设置道值")
    set_dao_level: int | None = Field(default=None, ge=1, description="直接设置道等级")
    lock_fate_dao: str | None = Field(
        default=None,
        description="跳过 roll 直接锁定本命道（联调）",
    )
    m6_quick_kit: bool | None = Field(
        default=None,
        description="一键：真仙+锁炎道+灌池+道等级3+道值500+开窗+清冷却+刷快照",
    )

    @model_validator(mode="after")
    def at_least_one_field(self) -> "GmSetCharacterRequest":
        """确保至少修改一项。"""
        values = [
            self.major_realm,
            self.realm_stage,
            self.cultivation_points,
            self.realm_progress,
            self.body_tempering_points,
            self.crafting_exp,
            self.breakthrough_grade,
            self.membership_tier,
            self.spirit_stones,
            self.idle_direction,
            self.status,
            self.clear_offline_pending,
            self.set_stamina,
            self.trial_puppet_count,
            self.reset_snapshot_cooldown,
            self.force_refresh_snapshot,
            self.divine_sense_capacity_bonus,
            self.array_craft_level,
            self.force_jindan,
            self.grant_craft_materials,
            self.grant_test_pet,
            self.clear_craft_jobs,
            self.clear_divine_sense_backlash,
            self.force_shichen,
            self.force_weather,
            self.start_tribulation,
            self.set_awaiting_ferry,
            self.force_ferry_timeout,
            self.mark_story_node,
            self.fate_luck,
            self.demonic_nature,
            self.force_yuanying_peak,
            self.spirit_root_tags,
            self.force_tribulation_outcome,
            self.grant_acceptance_constitution,
            self.force_true_immortal,
            self.grant_dao_pool,
            self.set_dao_lord,
            self.open_dao_challenge_window,
            self.open_dao_contest_now,
            self.clear_dao_challenge_cooldown,
            self.push_world_env,
            self.set_dao_qi,
            self.set_dao_level,
            self.lock_fate_dao,
            self.m6_quick_kit,
        ]
        if all(item is None for item in values):
            raise ValueError("至少提供一项要修改的字段")
        return self

File: frontend/src/api/gm.ts
/**
 * GM API：仅开发联调（M1～M6）；生产构建 UI 不展示入口。
 */
import { http } from './http'
import { envelopeFromAxiosError } from './envelope'
import type { ApiResponse } from '../types/api'
import type { CharacterPublic } from '../types/character'

/** GM 写入字段（皆可选） */
export interface GmSetCharacterPayload {
  major_realm?: string
  realm_stage?: number
  cultivation_points?: number
  realm_progress?: number
  body_tempering_points?: number
  crafting_exp?: number
  spirit_stones?: number
  idle_direction?: string
  status?: string
  breakthrough_grade?: string
  membership_tier?: string
  clear_offline_pending?: boolean
  // --- M3 战斗成型调试字段 ---
  /** 直接设置体力值 */
  set_stamina?: number
  /** 设置试炼木傀持有数 */
  trial_puppet_count?: number
  /** 清除快照手动更新冷却 */
  reset_snapshot_cooldown?: boolean
  /** 立即用当前防守预设重建快照 */
  force_refresh_snapshot?: boolean
  // --- M4 双线程调试字段 ---
  /** 一键提升至金丹（联调用） */
  force_jindan?: boolean
  /** 发放工坊材料样本 */
  grant_craft_materials?: boolean
  /** 发放测试灵宠 */
  grant_test_pet?: boolean
  /** 清空工坊队列 */
  clear_craft_jobs?: boolean
  /** 神识容量 GM 加成 */
  divine_sense_capacity_bonus?: number
  /** 清除神识反噬标记 */
  clear_divine_sense_backlash?: boolean
  /** 阵法钻研等级 */
  array_craft_level?: number
  // --- M5 环境与轮回调试 ---
  /** 强制当前时辰 */
  force_shichen?: string
  /** 强制天气 */
  force_weather?: string
  /** 灵根环境标签（参与 tag_modifiers；空数组清空） */
  spirit_root_tags?: string[]
  /** 强制渡劫结局：won / failed / fallen */
  force_tribulation_outcome?: string
  /** 发放验收体质（主/副词条）并自动镶嵌 */
  grant_acceptance_constitution?: boolean
  /** 一键开启渡劫会话 */
  start_tribulation?: boolean
  /** 置为待引渡 */
  set_awaiting_ferry?: boolean
  /** 立即触发待引渡超时轮回 */
  force_ferry_timeout?: boolean
  /** 标记剧情已历节点 */
  mark_story_node?: string
  /** 一键元婴大圆满 */
  force_yuanying_peak?: boolean
  // --- M6 大道 / 道主 ---
  /** 一键真仙初期（可开道） */
  force_true_immortal?: boolean
  /** 灌入道池 id 列表 */
  grant_dao_pool?: string[]
  /** 任命道主；空字符串清空自己的道主身份 */
  set_dao_lord?: string
  /** 强制挑战开窗（进程标志） */
  open_dao_challenge_window?: boolean
  /** M6-D06：立刻开赛（进程标志） */
  open_dao_contest_now?: boolean
  /** 清空挑战冷却 */
  clear_dao_challenge_cooldown?: boolean
  /** 经 WS 推一条 world.env */
  push_world_env?: boolean
  /** 直接设道值 */
  set_dao_qi?: number
  /** 直接设道等级 */
  set_dao_level?: number
  /** 跳过 roll 锁定本命道 */
  lock_fate_dao?: string
  /** 一键 M6 联调套装 */
  m6_quick_kit?: boolean
}

/**
 * POST /gm/character/set
 *
 * @param payload - 至少一项
 */
export async function gmSetCharacterApi(
  payload: GmSetCharacterPayload,
): Promise<ApiResponse<{ character: CharacterPublic }>> {
  try {
    const response = await http.post<ApiResponse<{ character: CharacterPublic }>>(
      '/gm/character/set',
      payload,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{ character: CharacterPublic }>(error)
  }
}

/** M5：强制时辰（经 character/set） */
export async function gmForceShichenApi(
  shichen: string,
): Promise<ApiResponse<{ character: CharacterPublic }>> {
  return gmSetCharacterApi({ force_shichen: shichen })
}

/** M5：强制天气 */
export async function gmForceWeatherApi(
  weather: string,
): Promise<ApiResponse<{ character: CharacterPublic }>> {
  return gmSetCharacterApi({ force_weather: weather })
}

/** M5：写入灵根环境标签 */
export async function gmSetSpiritRootTagsApi(
  tags: string[],
): Promise<ApiResponse<{ character: CharacterPublic }>> {
  return gmSetCharacterApi({ spirit_root_tags: tags })
}

/** M5：一键开渡劫 */
export async function gmStartTribulationApi(): Promise<
  ApiResponse<{ character: CharacterPublic }>
> {
  return gmSetCharacterApi({ start_tribulation: true })
}

/** M5：强制渡劫结局（验收） */
export async function gmForceTribulationOutcomeApi(
  outcome: 'won' | 'failed' | 'fallen',
): Promise<ApiResponse<{ character: CharacterPublic }>> {
  return gmSetCharacterApi({ force_tribulation_outcome: outcome })
}

/** M5：发放验收体质并自动镶嵌 */
export async function gmGrantAcceptanceConstitutionApi(): Promise<
  ApiResponse<{ character: CharacterPublic }>
> {
  return gmSetCharacterApi({ grant_acceptance_constitution: true })
}

/** M5：置待引渡 */
export async function gmSetAwaitingFerryApi(): Promise<
  ApiResponse<{ character: CharacterPublic }>
> {
  return gmSetCharacterApi({ set_awaiting_ferry: true })
}

/** M5：立即超时轮回 */
export async function gmForceFerryTimeoutApi(): Promise<
  ApiResponse<{ character: CharacterPublic }>
> {
  return gmSetCharacterApi({ force_ferry_timeout: true })
}

/** M5：标记 story 节点 */
export async function gmMarkStoryNodeApi(
  nodeId: string,
): Promise<ApiResponse<{ character: CharacterPublic }>> {
  return gmSetCharacterApi({ mark_story_node: nodeId })
}

/** M6：一键真仙 */
export async function gmForceTrueImmortalApi(): Promise<
  ApiResponse<{ character: CharacterPublic }>
> {
  return gmSetCharacterApi({ force_true_immortal: true })
}

/** M6：一键联调套装 */
export async function gmM6QuickKitApi(): Promise<
  ApiResponse<{ character: CharacterPublic }>
> {
  return gmSetCharacterApi({ m6_quick_kit: true })
}

/** M6：锁本命道 */
export async function gmLockFateDaoApi(
  daoId: string,
): Promise<ApiResponse<{ character: CharacterPublic }>> {
  return gmSetCharacterApi({ lock_fate_dao: daoId })
}

/** M6：设道资源 */
export async function gmSetDaoResourcesApi(opts: {
  qi?: number
  level?: number
}): Promise<ApiResponse<{ character: CharacterPublic }>> {
  return gmSetCharacterApi({
    set_dao_qi: opts.qi,
    set_dao_level: opts.level,
  })
}

/** M6：任命/清空道主 */
export async function gmSetDaoLordApi(
  daoId: string,
): Promise<ApiResponse<{ character: CharacterPublic }>> {
  return gmSetCharacterApi({ set_dao_lord: daoId })
}

/** M6：强制开窗 */
export async function gmOpenDaoChallengeWindowApi(): Promise<
  ApiResponse<{ character: CharacterPublic }>
> {
  return gmSetCharacterApi({ open_dao_challenge_window: true })
}

/** M6-D06 P1：立刻开赛 */
export async function gmOpenDaoContestNowApi(): Promise<
  ApiResponse<{ character: CharacterPublic }>
> {
  return gmSetCharacterApi({ open_dao_contest_now: true })
}

/** M6：推 world.env */
export async function gmPushWorldEnvApi(): Promise<
  ApiResponse<{ character: CharacterPublic }>
> {
  return gmSetCharacterApi({ push_world_env: true })
}
